private/protocol/rest/v2/encode.go

```diff
@@ -0,0 +1,75 @@
+package rest
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// An Encoder provides encoding of REST URI path, query, and header components
+// of an HTTP request. Can also encode a stream as the payload.
+//
+// Does not support SetFields.
+type Encoder struct {
+	req *http.Request
+
+	path, rawPath, pathBuffer []byte
+
+	query  url.Values
+	header http.Header
+}
+
+// NewEncoder creates a new encoder from the passed in request. All query and
+// header values will be added on top of the request's existing values. Overwriting
+// duplicate values.
+func NewEncoder(req *http.Request) *Encoder {
+	e := &Encoder{
+		req: req,
+
+		path:    []byte(req.URL.Path),
+		rawPath: []byte(req.URL.Path),
+		query:   req.URL.Query(),
+		header:  req.Header,
+	}
+
+	return e
+}
+
+// Encode will return the request and body if one was set. If the body
+// payload was not set the io.ReadSeeker will be nil.
+//
+// returns any error if one occured while encoding the API's parameters.
+func (e *Encoder) Encode() *http.Request {
+	e.req.URL.Path, e.req.URL.RawPath = string(e.path), string(e.rawPath)
+	e.req.URL.RawQuery = e.query.Encode()
+	e.req.Header = e.header
+
+	return e.req
+}
+
+func (e *Encoder) AddHeader(key string) *HeaderValue {
+	return newHeaderValue(e.header, key, true)
+}
+
+func (e *Encoder) SetHeader(key string) *HeaderValue {
+	return newHeaderValue(e.header, key, false)
+}
+
+func (e *Encoder) Headers(prefix string) *Headers {
+	return &Headers{
+		header: e.header,
+		prefix: strings.TrimSpace(prefix),
+	}
+}
```

Comment:
Can this builder be done without returning a new pointer for the `Headers` structure?  Limiting the pointer use to only where needed will help reduce allocations. This will increase the struct copies but the copies shouldn't need to be put onto the heap, should only exist on the stack.

Since the underlying `e.header` is a pointer we probably can get away with the value setter types not being pointers.